docs/tutorials/query-cache.txt

```diff
@@ -0,0 +1,213 @@
+***********
+Query Cache
+***********
+
+.. default-domain:: mongodb
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+.. _query-cache:
+
+The MongoDB Ruby driver provides a built-in query cache. When enabled, the
+query cache saves the results of previously-executed find and aggregation
+queries. When those same queries are performed again, the driver returns
+the cached reuslts to prevent unnecessary roundtrips to the database.
+
+Usage
+=====
+
+The query cache is disabled by default. It cache can be enabled on the global
+scope as well as within the context of a specific block.
+
+To enable the query cache globally:
+
+.. code-block:: ruby
+
+  Mongo::QueryCache.enabled = true
+
+Similarly, to disable it globally:
+
+.. code-block:: ruby
+
+  Mongo::QueryCache.enabled = false
+
+To enable the query cache within the context of a block:
+
+.. code-block:: ruby
+
+  Mongo::QueryCache.cache do
+    Mongo::Client.new([ '127.0.0.1:27017' ], database: 'music') do |client|
+      client['artists'].find(name: 'Flying Lotus').first
+      #=> Queries the database and caches the result
+
+      client['artists'].find(name: 'Flying Lotus').first
+      #=> Returns the previously cached result
+    end
+  end
+
+And to disable the query cache in the context of a block:
+
+.. code-block:: ruby
+
+  Mongo::QueryCache.uncached do
+    Mongo::Client.new([ '127.0.0.1:27017' ], database: 'music') do |client|
+      client['artists'].find(name: 'Flying Lotus').first
+      #=> Sends the query to the database; does NOT cache the result
+
+      client['artists'].find(name: 'Flying Lotus').first
+      #=> Queries the database again
+    end
+  end
+
+You may check whether the query cache is enabled at any time by calling
+``Mongo::QueryCache.enabled?``, which will return ``true`` or ``false``.
+
+.. _query-cache-matching:
+
+Query Matching
+==============
+
+A query is eligible to use cached results if it matches the original query
+that produced the cached results. Two queries are considered matching if they
+are identical in the following values:
+
+* Namespace (the database and collection on which the query was performed)
+* Selector (for aggregations, the aggregation pipeline stages)
+* Skip
+* Sort
+* Projection
+* Collation
+* Read Concern
+* Read Preference
+
+For example, if you perform one query, and then perform a mostly identical query
+with a different sort order, those queries will not be considered matching,
+and the second query will not use the cached results of the first.
```

Comment:
@p-mongo What do you think of this section?

Reply:
:+1: